docs/apache-airflow-providers/core-extensions/deprecations.rst

```diff
@@ -0,0 +1,57 @@
+ .. Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+ ..   http://www.apache.org/licenses/LICENSE-2.0
+
+ .. Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+
+Deprecations
+==========================================
+
+This is a summary of deprecated objects in Apache Airflow Providers Packages.
+
+.. note::
+   At the moment we only show deprecations for classes, functions, methods, properties etc.
+   Support for an argument deprecation or an argument value deprecation will be added in the future.
+
+
+.. important::
+   Only deprecations made with @deprecated decorator are included in this docs.
+   If You use warnings.warn in function/method body it will not show up in this documentation.
+   Make sure to use @deprecated decorator as shown in the below example:
```

Comment:
True, this is mostly for contributors or developers that are trying to see if the deprecation from their package is visible. Do You have any idea where can i put it? Or just leave it as a comment maybe?

Reply:
It should be in some contributor facing doc but I think it's better to find a way to enforce it rather than documenting it (for example: pre-commit that verify that all deprecations must be in the method you set as expected)
For example: We added pre-commit that verifies in providers we use only `AirflowProviderDeprecationWarning` so we may need another pre-commit to verify the method this `AirflowProviderDeprecationWarning` is being used.

https://github.com/apache/airflow/blob/48158c9967c14acf65264e338b8dea60af3f25e1/.pre-commit-config.yaml#L517-L521

Reply:
I agree with the idea of enforcement rather than documenting, but i think this one is tricky at the moment. I believe ideally we should verify if there are no `warnings.warn` with`AirflowProviderDeprecationWarning` being raised anywhere in the function/method body, only in decorators, but at the moment there is no decorator for deprecating argument or their values. So we can't really tell if the `warnings.warn` in `__init__` is about the class or some argument of init (without parsing the message and taking a guess). Do You have any idea how i could enforce it ?

Reply:
@eladkal Do You have any thoughts on that maybe ?

Reply:
Chiming in here. I think it's difficult to make the distinction now. but we should be able to influence the deprecations are raised.  For example we could define a new AirfllowProviderAttributeDeprecationWarning(AirflowProviderDeprecationWarning) - taking the name of attribute, and new_name of the atribute as parameters. And only allow to use those in the body of the method.

Reply:
I think the documentation on deprecations targetted for contributors, should find its way to the new contributing docs - likely here: https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#coding-style-and-best-practices

Reply:
@potiuk Yes, i believe we can standardize the way the deprecations are made. I like the idea of specific deprecation warnings, but, correct me if I'm wrong, if we put this new kind of Warning in the core Airflow ( f.e. `airflow/exceptions.py`) then the providers will not be able to use it unless they set the required minimum version of Airflow to f.e. 2.8.2 (the next one from now), which can never happen for some of them, therefore, it's going to be hard to rely on it in documentation? We can ofc try some try/except on import, but I'm not sure if it will solve the problem ?

My idea for standarization of deprecations is to use decorators #36952 . We could define decorators like: @deprecated, @deprecated_args, @deprecated_arg_values like in tensorflow [codebase](https://github.com/tensorflow/tensorflow/blob/dec8e0b11f4f87693b67e125e67dfbc68d26c205/tensorflow/python/util/deprecation.py#L274), we could decorate the exact function/class it touches and possibly eliminate the need of using warnings.warn within the body.

Anyway, maybe it will be good for this particular PR to remove this part of docs that we refer to here, and mark this doc as experimental or something similar? Then we can try to work on some rules, document (in the place You mentioned) and enforce them in a separate PR, f.e. the one i mentioned. Let me know what You think.

Reply:
Yes. You are right it's tricky for providers to use new features imported from common airlfow code. 
I thought about it and there is a way, though rather complex-ish - we could release another common provider  (`common.utils` for the lack of better name though I hate generic utils) where we could put such `never-backwards-incompatible` base classes that could be imported by providers. Similar to common.io. 

That would remove the dependency on airflow core and your https://github.com/apache/airflow/pull/36952 could land there, rather than in airflow core (or at least the part of it that is going to be used in providers).

And this is probably the only way we can implement decorators that we will be able to use in providers immediately (assuming that the `common.utils` will never be incompatible.

Maybe that's the right approach ?

Reply:
As mentioned, i removed the part of the docs about how the deprecation should be made - it will be enforced once we implement a standardized way of deprecating across airflow codebase.